Macos native statusbar

vslavik opened the pull request:
This PR removes generic code to mimic native macOS status bars, which we had to update in the past for new OS X versions (typically with a significant delay). Instead, it does what is apparently the "official" way of having status bar: set large bottom content border thickness, which makes the OS render the desired background, and draw status items on it.

Thus, we remove some code *and* get perfect match to the native look (with a Big Sur-sized caveat below). 

This PR depends on PR #2158, so has an extra commit in it (5ebd76c). The subject matter commits are only these:
- e687d0c - Render statusbar natively on macOS 
- d28771c - Tweak wxStatusBar size for macOS 11

Notice that this is only possible thanks to #2158. Without it, the background would be rendered as solid-color because wx would unconditionally set the background color on `wxFrame`.

Also notice that the Big Sur commit adjusts statusbar size to match Finder (as was done previously too; bottom border default in Interface Builder is only 22pt, not 24), but doesn't change rendering method. It uses the same native bottom border rendering, which Apple did not, as of macOS 11.1, update for the new visuals. Finder in macOS 11 does *not* use bottom border anymore, but renders the status bar same as path bar, in only the content side of the window. This means there's pretty much no presence of native statusbar in macOS now. It *also* means that it is entirely possible Apple won't fix bottom border rendering (because they discourage statusbars altogether) and consequently that there's an argument for rejecting this PR.

vslavik committed:
Fix window background tinting on macOS 11

macOS 11 has an option (on by default) to tint window backgrounds with wallpaper colors. This means that standard window background color is not a constant anymore and can change as the window is moved across the screen.

The key to supporting this is to _not set background color_ internally to what we think is the correct default color, or to repaint backgrounds. Let the OS handle the default behavior instead.

vslavik committed:
Render statusbar natively on macOS

Instead of trying to mimic native statusbar background rendering in our
own code, use NSWindow setContentBorderThickness:forEdge: to extend the
border that is drawn by the system. This is the canonical way of doing
statusbars ("bottom bar") in AppKit.

The text is still drawn generically, but that's a much smaller issue
than trying to match the border gradient.

vslavik committed:
Tweak wxStatusBar size for macOS 11

Finder's status bar size increased by 4pt in Big Sur.

vadz commented on src/osx/carbon/statbrma.cpp:
```diff
@@ -78,42 +78,22 @@ void wxStatusBarMac::InitColours()
-            m_bgActiveTo = wxColour(0x29, 0x29, 0x2A);
-            m_borderActive = wxColour(0x00, 0x00, 0x00);
-            m_borderInactive = wxColour(0x00, 0x00, 0x00);
+            m_textActive = wxColour(0xA9, 0xA9, 0xA9);
+            m_textInactive = wxColour(0x67, 0x67, 0x67);
```
Do we still have no way to get these colours from the system?

csomor replied:
I don't know which exactly, but we if we expose this (what would be the best way ? constants ? enums )

https://developer.apple.com/design/human-interface-guidelines/macos/visual-design/color#system-colors

we should be able to find a match

vslavik replied:
I don't think these colors are among system colors (no reason to: this is OS-rendered chrome), and certainly not with matching *semantics*. The only one semantically close is `windowFrameTextColor`, which doesn't match (and doesn't handle inactivated state).

vadz reviewed (approved):
Thanks, I don't have any real comments here, not drawing the background is (IMO) always better than drawing it, so I'm for applying this PR, of course. Please correct me if I'm wrong, but I don't think it makes things any worse under Big Sur, does it?

BTW, looking at the code it seems that `wxStatusBarMac` overrides a couple of methods which it might not have to override. Notably, `DrawField()` does the same thing as the base class version when `wxSB_NORMAL` (default) panes are used. `DrawFieldText()` contains a single Mac-specific test, but doesn't handle ellipsization, unlike the base class version, so it would be better to avoid overriding it too (is `wxFRAME_EX_METAL` still used?).

csomor reviewed (approved):
looks perfect, thanks a lot

vslavik commented:
> Please correct me if I'm wrong, but I don't think it makes things any worse under Big Sur, does it?

Correct.

I wrote that thinking it's going to make things marginally harder to implement again (i.e. we're now removing code that will partially need to be re-added back), but I don't agree with yesterday-me on that subject: it's probably just one `Clear()` call with the right color.

> `DrawFieldText()` contains a single Mac-specific test, but doesn't handle ellipsization, unlike the base class version, so it would be better to avoid overriding it too (is `wxFRAME_EX_METAL` still used?).

Nice catch. Apparently `wxFRAME_EX_METAL` is used for textured window backgrounds (deprecated in 10.14, and I am a bit unclear on its usefulness). But notably, the metal *appearance* doesn't exist from OS X 10.10 onwards. So this code can indeed be removed.

vadz commented:
> 
> 
> > `DrawFieldText()` contains a single Mac-specific test, but doesn't handle ellipsization, unlike the base class version, so it would be better to avoid overriding it too (is `wxFRAME_EX_METAL` still used?).
> 
> Nice catch. Apparently `wxFRAME_EX_METAL` is used for textured window backgrounds (deprecated in 10.14, and I am a bit unclear on its usefulness). But notably, the metal _appearance_ doesn't exist from OS X 10.10 onwards. So this code can indeed be removed.

Thanks for confirming this!

@csomor I propose to remove all tests for `wxFRAME_EX_METAL` and define it as 0. Do you have any objections?

vslavik commented:
> @csomor I propose to remove all tests for wxFRAME_EX_METAL and define it as 0. Do you have any objections?

To be clear, `wxFRAME_EX_METAL` now maps to `NSWindowStyleMaskTexturedBackground` and that does *something* (I think, ability to drag the window anywhere in its area). It just doesn't have anything to do with "metal", doesn't affect appearance, but is not a no-op.

It is certainly safe to remove its `wxStatusBar` use specifically.

csomor commented:
> > Nice catch. Apparently `wxFRAME_EX_METAL` is used for textured window backgrounds (deprecated in 10.14, and I am a bit unclear on its usefulness). But notably, the metal _appearance_ doesn't exist from OS X 10.10 onwards. So this code can indeed be removed.
> 
> Thanks for confirming this!
> 
> @csomor I propose to remove all tests for `wxFRAME_EX_METAL` and define it as 0. Do you have any objections?

the metal appearance is long gone, but a textured window still looks different if you eg create one with a toolbar in interface builder, the content area being darker of course but even the toolbar has a slightly different gradient. Also the dragging behaviour is different. A textured window can be dragged by clicking anywhere in the background.

vslavik committed:
Remove redundant code from wxStatusBarMac

Remove a lot of Mac-specific code from wxStatusBarMac in favour of using
shared wxStatusBarGeneric:

- Now that wxFRAME_EX_METAL aka NSWindowStyleMaskTexturedBackground
  doesn’t affect text placement, there’s no reason for customized
  DrawFieldText(). As a side effect of this removal, ellipsizing fields
  is now supported.

- Remove customized DoUpdateStatusText() code that is no longer needed.

See https://github.com/wxWidgets/wxWidgets/pull/2160#pullrequestreview-563916013

vslavik commented:
I pushed the cleanup as 634f60ff. I am reasonably sure of it; the key word being "reasonably". `DoUpdateStatusText()` had  some strange differences that this removes:
- it used taller refresh rect (entire statusbar width vs just the field)
- it had commented-out call to `Update()`

That commented-out change is 7.5 years old and only explained as "avoiding reentrancy problems and congestion" (5856534c8683010260628adc346600b796cffb02). I think it's a reasonable bet that it no longer applies. The strange refresh rect is even older and probably related to some rendering issues not present on modern macOS either.
